File: tech-studio-projects/financial-industry/crypto-trading-simulation/java/src/core/order_router/OrderRouter.cpp
#include "OrderRouter.h"
#include <curl/curl.h>
#include "../utils/Config.h"
#include "../utils/Logger.h"
#include "../utils/HmacSha256.h"

void OrderRouter::placeOrder(const std::string& symbol, const std::string& side, double quantity, double price) {
    CURL* curl = curl_easy_init();
    if(curl) {
        // Load API keys from config.json
        std::string api_key = Config::get("BINANCE_API_KEY");
        std::string api_secret = Config::get("BINANCE_API_SECRET");

        if (api_key.empty() || api_secret.empty()) {
            Logger::logError("API keys are missing in config");
            curl_easy_cleanup(curl);
            return;
        }

        // Construct the order payload
        std::string query = "symbol=" + symbol + "&side=" + side + "&type=LIMIT&timeInForce=GTC";
        query += "&quantity=" + std::to_string(quantity) + "&price=" + std::to_string(price);
        query += "&timestamp=" + std::to_string(time(NULL) * 1000);

        // Sign the request
        std::string signature = hmac_sha256(api_secret, query);
        query += "&signature=" + signature;

        // Set the API endpoint and headers
        std::string url = "https://api.binance.com/api/v3/order";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, query.c_str());
        struct curl_slist* headers = NULL;
        headers = curl_slist_append(headers, ("X-MBX-APIKEY: " + api_key).c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

        // Perform the request and handle response
        CURLcode res = curl_easy_perform(curl);
        if(res != CURLE_OK) {
            Logger::logError("Failed to place order: " + std::string(curl_easy_strerror(res)));
        } else {
            Logger::logInfo("Order placed successfully.");
        }
        
        curl_easy_cleanup(curl);
    }
}

File: tech-studio-projects/financial-industry/cryptocurrency-trading-simulation/java/src/core/order_router/OrderRouter.java
// OrderRouter.java
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.util.Map;
import java.time.Instant;

public class OrderRouter {

    // Method to place an order
    public void placeOrder(String symbol, String side, double quantity, double price) {
        try {
            // Load API keys from Config class (assuming a similar utility class exists in Java)
            String apiKey = Config.get("BINANCE_API_KEY");
            String apiSecret = Config.get("BINANCE_API_SECRET");

            if (apiKey == null || apiSecret == null || apiKey.isEmpty() || apiSecret.isEmpty()) {
                Logger.logError("API keys are missing in config");
                return;
            }

            // Construct the order payload
            String query = "symbol=" + symbol + "&side=" + side + "&type=LIMIT&timeInForce=GTC" +
                           "&quantity=" + quantity + "&price=" + price +
                           "&timestamp=" + Instant.now().toEpochMilli();

            // Sign the request
            String signature = hmacSha256(apiSecret, query);
            query += "&signature=" + signature;

            // Set the API endpoint and headers
            String url = "https://api.binance.com/api/v3/order";
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("X-MBX-APIKEY", apiKey)
                .POST(HttpRequest.BodyPublishers.ofString(query))
                .build();

            // Send the request and handle the response
            HttpClient client = HttpClient.newHttpClient();
            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                Logger.logError("Failed to place order: " + response.body());
            } else {
                Logger.logInfo("Order placed successfully.");
            }
        } catch (Exception e) {
            Logger.logError("An error occurred: " + e.getMessage());
        }
    }

    // HMAC-SHA256 signing function
    private String hmacSha256(String key, String data) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKeySpec secretKeySpec = new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        mac.init(secretKeySpec);
        byte[] hash = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
        StringBuilder hexString = new StringBuilder();
        for (byte b : hash) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) hexString.append('0');
            hexString.append(hex);
        }
        return hexString.toString();
    }
}
